composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,221 @@
+    # |DWA result (float)|...                |DWA result (float)|
+    # +------------------+-------------------+------------------+
+    phx_annual_prcp_df = spark.createDataFrame([[]], StructType([]))
+    phx_annual_snow_df = spark.createDataFrame([[]], StructType([]))
+
+    # distance weighting algorithm (DWA)
+    def phx_dw_compute(input_list) -> float:
+        # input_list is a list of Row object with format:
+        # [
+        #    ROW(ID='...', LATITUDE='...', LONGITUDE='...', YEAR='...', ANNUAL_PRCP/ANNUAL_SNOW='...'),
+        #    ROW(ID='...', LATITUDE='...', LONGITUDE='...', YEAR='...', ANNUAL_PRCP/ANNUAL_SNOW='...'),
+        #    ...
+        # ]
+
+        # contains 1 / d^2 of each station
+        factor_list = []
+        # the total sum of 1 / d^2 of all the stations
+        factor_sum = 0.0
+        for row in input_list:
+            latitude = row[1]
+            longitude = row[2]
+            # calculate the distance from each station to Phoenix
+            distance_to_phx = math.sqrt((phx_location[0] - latitude) ** 2 + (phx_location[1] - longitude) ** 2)
+            # calculate the distance factor of each station (1 / d^2)
+            distance_factor = 1.0 / (distance_to_phx ** 2)
+            factor_list.append(distance_factor)
+            factor_sum += distance_factor
+
+        # contains the weights of each station
+        weights_list = []
+        for val in factor_list:
+            weights_list.append(val / factor_sum)
+
+        dwa_result = 0.0
+        for row in input_list:
+            # this is the annual prcipitation/snowfall of each station
+            annual_value = row[4]
+            # weight of each station
+            weight = weights_list[input_list.index(row)]
+            dwa_result += weight * annual_value
+
+        print("weight list", weights_list)
+        print("factor list", factor_list)
+        print("return result", dwa_result)
+        print("factor sum", factor_sum)
+
+        return dwa_result
+
+    for year in range(1997, 2022):
+        # collect() function returns a list of Row object.
+        # prcp_year and snow_year will be the input for the distance weighting algorithm
+        prcp_year = (
+            annual_df.where((annual_df.ELEMENT == 'PRCP') & (annual_df.YEAR == year))
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_PRCP")).collect()
+        )
+        # print(prcp_year)
+        snow_year = (
+            annual_df.where((annual_df.ELEMENT == 'SNOW') & (annual_df.YEAR == year))
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_SNOW")).collect()
+        )
+
+        phx_annual_prcp_df = (
+            phx_annual_prcp_df.withColumn(f"PHX_PRCP_{year}", lit(phx_dw_compute(prcp_year)))
+        )
+        phx_annual_snow_df = (
+            phx_annual_snow_df.withColumn(f"PHX_SNOW_{year}", lit(phx_dw_compute(snow_year)))
+        )
```

Comment:
You can also just create the DataFrames here and populate them in one line instead of declaring them on 119 / 120. See Section 2.2: https://sparkbyexamples.com/pyspark/different-ways-to-create-dataframe-in-pyspark/

Reply:
The reason I'm doing this is because they are in a `for loop`. If I didn't misunderstand your point, I think creating and populating them in one line would result in more than one DF, which is not what I want.